### ISSUE TITLE
warnings stopping build (e999aa4)

### ISSUE THREAD

**Issue by Digital-Monk** [colah/ImplicitCAD#244]
(I'm only filing this because I followed the build instructions and it's from a fresh clone, so it seems like it might be an issue of something missing from the repo?  If not, I apologize profusely.)

Cloned the repo, installed dependencies, configured and tried to install.  `cabal install` dumped this:

```
cabal: Entering directory '.'
Configuring implicit-0.2.1...
Building implicit-0.2.1...
Preprocessing library implicit-0.2.1...

<no location info>: error: 
Failing due to -Werror.

on the commandline: warning:
    unrecognised warning flag: -Wmissing-export-lists

on the commandline: warning:
    unrecognised warning flag: -Wmissing-home-modules
cabal: Leaving directory '.'
```

Guessing this is caused by a missing tool or library on my end...  I do have 0.2.0 installed as well, but that shouldn't cause compilation warnings.

Just to try to proceed, I went through implicit.cabal and commented out all of the -Werror options.  `cabal install` then generated a much longer list of issues:
```
Preprocessing library implicit-0.2.1...

on the commandline: warning:
    unrecognised warning flag: -Wmissing-export-lists

on the commandline: warning:
    unrecognised warning flag: -Wmissing-home-modules
[ 1 of 48] Compiling Graphics.Implicit.FastIntUtil ( Graphics/Implicit/FastIntUtil.hs, dist/build/Graphics/Implicit/FastIntUtil.o )
[ 2 of 48] Compiling Graphics.Implicit.IntegralUtil ( Graphics/Implicit/IntegralUtil.hs, dist/build/Graphics/Implicit/IntegralUtil.o )
[ 3 of 48] Compiling Graphics.Implicit.ExtOpenScad.Parser.Lexer ( Graphics/Implicit/ExtOpenScad/Parser/Lexer.hs, dist/build/Graphics/Implicit/ExtOpenScad/Parser/Lexer.o )

./Graphics/Implicit/ExtOpenScad/Parser/Lexer.hs: warning:
    Could not specialise imported function ‘char’
    Probable fix: add INLINEABLE pragma on ‘char’
[ 4 of 48] Compiling Graphics.Implicit.Definitions ( Graphics/Implicit/Definitions.hs, dist/build/Graphics/Implicit/Definitions.o )

./Graphics/Implicit/Definitions.hs: warning:
    Could not specialise imported function ‘Data.Either.$w$cshowsPrec’
      when specialising ‘Data.Either.$fShowEither_$cshowsPrec’
    Probable fix: add INLINEABLE pragma on ‘Data.Either.$w$cshowsPrec’

./Graphics/Implicit/Definitions.hs: warning:
    Could not specialise imported function ‘GHC.Show.$fShowMaybe_$cshowsPrec’
    Probable fix: add INLINEABLE pragma on ‘GHC.Show.$fShowMaybe_$cshowsPrec’

./Graphics/Implicit/Definitions.hs: warning:
    Could not specialise imported function ‘GHC.Show.$w$cshowsPrec4’
      when specialising ‘GHC.Show.$fShow(,,)_$cshowsPrec’
    Probable fix: add INLINEABLE pragma on ‘GHC.Show.$w$cshowsPrec4’
[ 5 of 48] Compiling Graphics.Implicit.Export.MarchingSquaresFill ( Graphics/Implicit/Export/MarchingSquaresFill.hs, dist/build/Graphics/Implicit/Export/MarchingSquaresFill.o )

Graphics/Implicit/Export/MarchingSquaresFill.hs:31:9: warning: [-Wmonomorphism-restriction]
    • The Monomorphism Restriction applies to the bindings for ‘n’,
                                                               ‘nx’, ‘ny’
        Consider giving a type signature for these binders
    • In the expression:
        let
          d = p2 ^-^ p1
          nx :: ℕ
          ny :: ℕ
          ....
        in triangles
      In an equation for ‘getContourMesh’:
          getContourMesh p1 p2 res obj
            = let
                d = p2 ^-^ p1
                nx :: ℕ
                ....
              in triangles

./Graphics/Implicit/Export/MarchingSquaresFill.hs: warning:
    Could not specialise imported function ‘Data.Foldable.$fFoldable[]_$cfold’
    Probable fix: add INLINEABLE pragma on ‘Data.Foldable.$fFoldable[]_$cfold’

./Graphics/Implicit/Export/MarchingSquaresFill.hs: warning:
    Could not specialise imported function ‘GHC.Float.$w$cproperFraction’
      when specialising ‘GHC.Float.$fRealFracDouble_$cproperFraction’
      when specialising ‘GHC.Float.$fRealFracDouble_$cfloor’
    Probable fix: add INLINEABLE pragma on ‘GHC.Float.$w$cproperFraction’
[ 6 of 48] Compiling Graphics.Implicit.Export.RayTrace ( Graphics/Implicit/Export/RayTrace.hs, dist/build/Graphics/Implicit/Export/RayTrace.o )

./Graphics/Implicit/Export/RayTrace.hs: warning:
    Could not specialise imported function ‘Data.VectorSpace.$w$cp1InnerSpace1’
      when specialising ‘Data.VectorSpace.$fInnerSpace(,,)_$cp1InnerSpace’
    Probable fix: add INLINEABLE pragma on ‘Data.VectorSpace.$w$cp1InnerSpace1’

./Graphics/Implicit/Export/RayTrace.hs: warning:
    Could not specialise imported function ‘GHC.Real.$w$s$cround’
      when specialising ‘GHC.Real.$fRealFracRatio_$s$cround’
    Probable fix: add INLINEABLE pragma on ‘GHC.Real.$w$s$cround’

./Graphics/Implicit/Export/RayTrace.hs: warning:
    Could not specialise imported function ‘Data.VectorSpace.$w$c<.>1’
      when specialising ‘Data.VectorSpace.$fInnerSpace(,,)_$c<.>’
      when specialising ‘Data.VectorSpace.$fInnerSpace(,,)’
    Probable fix: add INLINEABLE pragma on ‘Data.VectorSpace.$w$c<.>1’

./Graphics/Implicit/Export/RayTrace.hs: warning:
    Could not specialise imported function ‘GHC.Show.$w$cshowsPrec4’
      when specialising ‘GHC.Show.$fShow(,,)_$cshowsPrec’
    Probable fix: add INLINEABLE pragma on ‘GHC.Show.$w$cshowsPrec4’
[ 7 of 48] Compiling Graphics.Implicit.Export.Render.Definitions ( Graphics/Implicit/Export/Render/Definitions.hs, dist/build/Graphics/Implicit/Export/Render/Definitions.o )
[ 8 of 48] Compiling Graphics.Implicit.Export.Render.GetLoops ( Graphics/Implicit/Export/Render/GetLoops.hs, dist/build/Graphics/Implicit/Export/Render/GetLoops.o )

Graphics/Implicit/Export/Render/GetLoops.hs:8:72: error:
    Module ‘Prelude’ does not export ‘(<>)’
cabal: Leaving directory '.'
cabal: Error: some packages failed to install:
implicit-0.2.1 failed during the building phase. The exception was:
ExitFailure 1
```

**Comment by julialongtin** [comment]
What version of haskell is this? We've been testing on 8.4 and 8.6..

**Comment by Digital-Monk** [comment]
Oh, FFS...  GHC 8.0.2...

I'm on Linux Mint 19.3 (newest) which is based on Ubuntu 18.04 (new-ish, and latest LTS)...  You'd think that its ghc package might not be ancient...

Huh.  Went to GHC home, and the only Debian-based instructions are to get it from the repos, which is what left me stuck with 8.0.2  (well, technically the GHC website instructions say to get the ghc6 package, which suggests to me how stale that page may be)

I see that GHC is up to 8.8.2, just not sure how much of a headache it's going to be to get it.  Not especially looking forward to a source build and possible dependency nightmare.

Debian packages says that their stable repo has 8.4.4.  Ubuntu package search shows that they've limited it to 8.0.2 in bionic (18.04 LTS).  Time for me to get used to adjusting my repo sources to something newer and hopefully not hose my dependencies ;)

So, once again, this is nothing on your project at all.  May want to tweak your install instructions, because unfortunately `sudo apt install ghc` can be way behind on common distributions...

**Comment by Digital-Monk** [comment]
For anyone else having similar problems, the "Generic Linux" installer seems pretty smooth:  [https://www.haskell.org/platform/linux.html#linux-generic](url).  Gave me GHC 8.6.5 and cabal-install 3.0.0.0.  It also updated my command path, saving a minor step.

Followed the `git clone ...` and `cabal ...` instructions from here, and only had two issues:
1. Needed to specify an installdir.  I used /home/myname/.cabal/bin, because that was where the old stable release put itself.
2. Because I had the old stable release there, I also needed to overwrite old versions.

So, my final install command was:
`cabal install --installdir=/home/myname/.cabal/bin --overwrite-policy=always`

Your mileage may vary...
